src/tigerbeetle/inspect.zig

```diff
@@ -0,0 +1,1100 @@
+                    ),
+                    std.math.order(table_a.snapshot_min, table_b.snapshot_min),
+                    std.math.order(table_a.snapshot_max, table_b.snapshot_max),
+                    std.math.order(table_a.checksum, table_b.checksum),
+                }) |order| {
+                    if (order != .eq) return order == .lt;
+                }
+                // This *should* be unreachable, especially given the checksum comparison.
+                return false;
+            }
+        }.less_than);
+
+        inline for (StateMachine.Forest.tree_infos) |tree_info| {
+            if (tree_info.tree_id == filter.tree_id) {
+                for (tables_filtered.items) |*table| {
+                    try print_table_info(output, tree_info, table);
+                }
+                break;
+            }
+        } else {
+            try output.print("error: unknown tree_id={}\n", .{filter.tree_id});
+        }
+    }
+
+    fn read_buffer(
+        inspector: *Inspector,
+        zone: vsr.Zone,
+        offset_in_zone: u64,
+        comptime size: usize,
+    ) !*align(constants.sector_size) const [size]u8 {
+        const buffer = try inspector.allocator.alignedAlloc(u8, constants.sector_size, size);
+        errdefer inspector.allocator.free(buffer);
+
+        inspector.storage.read_sectors(
+            inspector_read_callback,
+            &inspector.read,
+            buffer,
+            zone,
+            offset_in_zone,
+        );
+        try inspector.work();
+        return buffer[0..size];
+    }
+
+    fn read_superblock(inspector: *Inspector, superblock_copy: u8) !SuperBlockHeader {
+        const superblock_buffer = try inspector.read_buffer(
+            .superblock,
+            @as(u64, superblock_copy) * vsr.superblock.superblock_copy_size,
+            @sizeOf(SuperBlockHeader),
+        );
+        defer inspector.allocator.free(superblock_buffer);
+
+        const superblock = std.mem.bytesAsValue(SuperBlockHeader, superblock_buffer);
+        return superblock.*;
+    }
+
+    fn read_block(inspector: *Inspector, address: u64, checksum: ?u128) !BlockPtrConst {
+        const buffer = try inspector.read_buffer(
+            .grid,
+            (address - 1) * constants.block_size,
+            constants.block_size,
+        );
+        errdefer inspector.allocator.free(buffer);
+
+        const header = std.mem.bytesAsValue(vsr.Header.Block, buffer[0..@sizeOf(vsr.Header)]);
+        if (!header.valid_checksum()) {
+            log.err(
+                "read_block: invalid block address={} checksum={?x:0>32} (bad checksum)",
+                .{ address, checksum },
+            );
+            return error.InvalidChecksum;
+        }
+
+        if (!header.valid_checksum_body(buffer[@sizeOf(vsr.Header)..header.size])) {
+            log.err(
+                "read_block: invalid block address={} checksum={?x:0>32} (bad checksum_body)",
+                .{ address, checksum },
```

Comment:
Feels worthwhile to print the checksum the block actually has as well, if only to make the output unambigious.